server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperation.java

```diff
@@ -68,24 +67,36 @@ public DelayedOperation(long delayMs, Lock lock) {
      * Return true iff the operation is completed by the caller: note that
      * concurrent threads can try to complete the same operation, but only
      * the first thread will succeed in completing the operation and return
-     * true, others will still return false
+     * true, others will still return false.
      */
     public boolean forceComplete() {
-        if (completed.compareAndSet(false, true)) {
-            // cancel the timeout timer
-            cancel();
-            onComplete();
-            return true;
-        } else {
+        // Do not proceed if the operation is already completed.
+        if (completed) {
```

Comment:
An implication of this change is that when forceComplete() is called inside `tryComplete`, it always returns true. Currently, we have multiple places that check the return value of forceComplete() inside `tryComplete`. Should we adjust those?

Reply:
@junrao that is a good point. After avoiding concurrent execution, `forceComplete` MUST return true if it is executed by `tryComplete`. Hence, checking the returned value of `forceComplete` gets meaningless.

Reply:
> Ok, so are we saying that first execution of forceComplete from tryComplete will return true? And as the operation is marked completed hence no further execution of tryComplete can trigger, which makes all executions of forceComplete within tryComplete always returns true?

yes, we are on the same page. I hope I'm not misunderstanding @junrao comment :)